### PR TITLE
Update functions and add functions to create target for masked patch prediction

### PR DESCRIPTION
**Pull request by boru-roylu**

### CODE REVIEW

File: src/input_utils.py [hunk 1]
```diff
@@ -56,14 +61,37 @@ class PretrainInputConfig(object):
   # Maximum number of masked patch tokens per batch.
   mpp_max_selections_per_batch = attr.ib(default=1600)
 
+  # Maximum output channel bits in masked patch prediction.
+  output_channel_bits = attr.ib(default=3)
+
```
Comment (wqfcr-google) [inline src/input_utils.py:67]: What do you mean by 'channel bits'?

File: src/input_utils.py [hunk 2]
```diff
@@ -56,14 +61,37 @@ class PretrainInputConfig(object):
...
+      return {k: v for k, v in self.__dict__.items() if v is not None}
+    else:
+      return dict(self.__dict__)
+
+
+def get_pretrain_example_decode_fn(
+  tokenizer: tf_text.BertTokenizer,
+  nput_config: PretrainInputConfig,
+  odel_config: EtcConfig,
+  s_training: bool):
```
Comment (wqfcr-google) [inline src/input_utils.py]: The parameter names seem broken. Please fix them.

File: src/input_utils.py [hunk 3]
```diff
@@ -149,7 +177,7 @@ def _decode_fn(record):
       begin, size, _ = tf.image.sample_distorted_bounding_box(
           tf.shape(im),
           tf.zeros([0, 0, 4], tf.float32),
-          area_range=(0.05, 1.0),
+          area_range=(0.85, 1.0),
```
Comment (wqfcr-google) [inline src/input_utils.py:180]: FYI: for the first version we don't need to worry about bounding box, but you can leave it as is for now.

File: src/pretrain_utils.py [hunk 4]
```diff
@@ -54,37 +63,101 @@ def input_fn_builder(tokenizer: tf_text.BertTokenizer,
...
+    Returns:
+      <tf.Tensor>[batch, sequence_length]
+    """
 
-  def make_masked_lm_features(features: Mapping[str, tf.Tensor]):
-    """Randomly mask out tokens
+    channels = get_shape_list(patch_tokens)[-1] // (patch_size**2)
+    bin_size = max_pixel_val // (2 ** output_channel_bits)
+
+    patch_tokens = patch_tokens * (max_pixel_val-1)
```
Comment (wqfcr-google) [inline src/pretrain_utils.py]: Nit: extra space around '-', same in other places.

File: src/pretrain_utils.py [hunk 5]
```diff
@@ -54,37 +63,101 @@ def input_fn_builder(tokenizer: tf_text.BertTokenizer,
...
+      patch_tokens: <tf.Tensor>[batch, sequence_length, patch_token_size]
+
+    Returns:
+      <tf.Tensor>[batch, sequence_length]
+    """
 
-  def make_masked_lm_features(features: Mapping[str, tf.Tensor]):
-    """Randomly mask out tokens
+    channels = get_shape_list(patch_tokens)[-1] // (patch_size**2)
+    bin_size = max_pixel_val // (2 ** output_channel_bits)
```
Comment (wqfcr-google) [inline src/pretrain_utils.py:81]: Maybe we could go over the code a bit together to help me understand the details.

File: src/input_utils.py (hunk 1 above)
Comment (boru-roylu) [reply src/input_utils.py:67]: 3-bit channel means we use 0-7 to represent the channel. A normally image has 8-bit channels. Instead of predicting 8-bit channels, we predict 3-bit channels for masked patch prediction. By doing so, we can make number of output classes smaller (2^3)^3=512 classes.  I follow the ViT paper and [PyTorch implementation](https://github.com/lucidrains/vit-pytorch/blob/22da26fa4b7d98c2a936b48c1b1dcf83c404dfb8/vit_pytorch/mpp.py#L33).

Comment (wqfcr-google) [reply src/input_utils.py:67]: I see. Interesting.

File: src/pretrain_utils.py [hunk 6]
```diff
@@ -54,37 +63,101 @@ def input_fn_builder(tokenizer: tf_text.BertTokenizer,
...
+    features['masked_text_target'] = txt_masked_ids
+      
     # Offset text positions after image ones ([CLS] [PATCH] P1 p2 ... P196).
-    text_positions = text_features[1] + 2 + num_patch_per_row ** 2
-    masked_input_ids = tf.concat([image_features[0], text_features[0]], axis=0)
-    masked_positions = tf.concat([image_features[1], text_positions], axis=0)
-    masked_ids = tf.concat([image_features[2], text_features[2]], axis=0)
+    txt_masked_positions = txt_masked_positions + 2 + num_patch_per_row ** 2
+
+    # Join
```
Comment (wqfcr-google) [inline src/pretrain_utils.py:157]: Nit: it is better to make the comment more meaningful:) For example, "Join image and text input features...."

**Review by wqfcr-google — COMMENTED** [review]
Overall looks good.

File: src/input_utils.py [hunk 7]
```diff
@@ -12,14 +12,17 @@
...
 import attr
 import tensorflow as tf
 import tensorflow_text as tf_text
+from official.modeling import tf_utils
 
-from google_research.etcmodel.models.modeling import EtcConfig 
 from tensor_utils import full, ragged_full
+from etcmodel import tensor_utils as etc_tensor_utils
+from etcmodel import feature_utils as etc_feature_utils
+from etcmodel.models.modeling import EtcConfig
```
Comment (jtainslie) [inline src/input_utils.py:25]: Nit: per the style guide, imports should be for modules and packages, not individual classes.  (https://google.github.io/styleguide/pyguide.html#22-importshttps://google.github.io/styleguide/pyguide.html#22-imports)

File: src/input_utils.py [hunk 8]
```diff
@@ -56,10 +62,32 @@ class PretrainInputConfig(object):
...
+  """RelativeTransformer side inputs ("att_mask" and "relative_att_ids").
+
+  See `RelativeTransformerLayers.call()` for a description of these side
+  inputs.
+  """
+
+  att_mask = attr.ib()  # type: Optional[tf.Tensor]
+  relative_att_ids = attr.ib()  # type: Optional[tf.Tensor]
+
+  def to_dict(self, exclude_none_values=False):
```
Comment (jtainslie) [inline src/input_utils.py:83]: Optional: it may be more convenient to set the default to `exclude_none_values=True` in new code like this.

File: src/pretrain_utils.py [hunk 9]
```diff
@@ -12,14 +12,20 @@
...
 import tensorflow_text as tf_text
+from tensorflow.python.ops import math_ops
+from official.modeling import tf_utils
 
-from google_research.etcmodel.models.modeling import EtcConfig 
-from input_utils import PretrainInputConfig, get_pretrain_example_decode_fn
+from etcmodel.models.modeling import EtcConfig 
+from input_utils import (PretrainInputConfig,
+                         get_pretrain_example_decode_fn,
+                         add_side_input_features)
```
Comment (jtainslie) [inline src/pretrain_utils.py:28]: Nit: imports should be for modules and packages only.  (https://google.github.io/styleguide/pyguide.html#22-importshttps://google.github.io/styleguide/pyguide.html#22-imports)

File: src/input_utils.py [hunk 10]
```diff
@@ -193,21 +222,97 @@ def _decode_fn(record):
...
+    name: A name for the operation (optional).
+
+  Returns:
+    A `RelativeTransformerSideInputs` with all relevant tensors set.
+  """
+
+  with tf.name_scope(name or 'make_relative_transformer_side_inputs'):
+    long_breakpoints = tf.convert_to_tensor(long_breakpoints)
+    long_example_ids = tf.cumsum(long_breakpoints, axis=-1, reverse=True)
+    long_example_ids = tf.convert_to_tensor(long_example_ids)
```
Comment (jtainslie) [inline src/input_utils.py:254]: Seems like this line should be unnecessary since `long_breakpoints` was already converted to a Tensor?

File: src/input_utils.py [hunk 11]
```diff
@@ -193,21 +222,97 @@ def _decode_fn(record):
...
+    long_seq_len = tf_utils.get_shape_list(long_example_ids)[1]
+    att_mask = etc_feature_utils.make_segmented_att_mask(long_example_ids)
+    batch_size = tf.shape(long_example_ids)[0]
```
Comment (jtainslie) [inline src/input_utils.py:257]: Nit: wouldn't hurt to use `get_shape_list` here too I think.

File: src/input_utils.py [hunk 12]
```diff
@@ -193,21 +222,97 @@ def _decode_fn(record):
...
+
+  Returns:
+    A new `features` dictionary with side inputs.
+  """
+
+  features = dict(features)
+
+  img_wp = features['num_image_wordpieces']
+  txt_wp = features['num_text_wordpieces']
+  seq_len = tf.math.add(img_wp, txt_wp)
```
Comment (jtainslie) [inline src/input_utils.py]: Nit: would this suffice?
```suggestion
  seq_len = img_wp + txt_wp
```

File: src/input_utils.py [hunk 13]
```diff
@@ -193,21 +222,97 @@ def _decode_fn(record):
...
+  max_seq_len_in_batch = tf.reduce_max(seq_len)
+  batch_size = tf_utils.get_shape_list(img_wp)[0]
+
+  img_wp = img_wp[:, tf.newaxis]
+  txt_wp = txt_wp[:, tf.newaxis]
+  helper = tf.range(max_seq_len_in_batch, dtype=tf.int32)
```
Comment (jtainslie) [inline src/input_utils.py:299]: Nit: a name like `position` may be more descriptive than `helper`.

**Review by wqfcr-google — APPROVED** [review]
Thanks Roy for the code walk through today!

**Review by jtainslie — APPROVED** [review]
Looks good apart from some nits.
